hamt/hamt.go

```diff
@@ -391,10 +392,16 @@ func (ds *Shard) getValue(ctx context.Context, hv *hashBits, key string, cb func
 // EnumLinks collects all links in the Shard.
 func (ds *Shard) EnumLinks(ctx context.Context) ([]*ipld.Link, error) {
 	var links []*ipld.Link
-	err := ds.ForEachLink(ctx, func(l *ipld.Link) error {
+	var setlk sync.Mutex
+
+	getLinks := ds.makeAsyncTrieGetLinks(func(l *ipld.Link) error {
+		setlk.Lock()
 		links = append(links, l)
+		setlk.Unlock()
 		return nil
 	})
+
+	err := dag.EnumerateChildrenAsync(ctx, getLinks, ds.nd.Cid(), func(c cid.Cid) bool { return true })
```

Comment:
The standard thing to pass as the visit function is a cid.Set's 'Visit' function.

Reply:
@whyrusleeping Does that make sense in this case? Given that the act of walking the child is what produces the shard output, I feel like we want a universal true here -- i.e. even if for some reason there were identical CIDs for two seperate shards, we will want to walk both shards so their directory output gets added to the overall list?

At the same time I'm having trouble imagining how this would be the case for a directory. Still I can definitely add it.

Reply:
I went ahead and added this: https://github.com/ipfs/go-unixfs/pull/19/commits/3738990784c63b9bfa2fc104c06c50ece6729aef

Didn't break tests, so I have no problem with having it here